Vault: Add entrypoint subpackage

amouat opened the pull request:
Should allow us to create an image that is close to upstream and usable as Helm replacement.

Also made symlink of vault binary to /usr/bin/, so we can later remove it from /usr/sbin

amouat committed:
Add entrypoint package to vault.

Should allow us to create an image that is close to upstream and usable
as Helm replacement.

Also made symlink of vault binary to /usr/bin/, so we can later remove
it from /usr/sbin

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat committed:
Fix YAM.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat commented:
One moment, forgot to bump epoch and want to check the tarfile...

amouat committed:
Bump epoch & fix symlink.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

tuananh commented:
i think the reason i put it in sbin is because I ported it from alpine package.

https://git.alpinelinux.org/aports/tree/community/vault/APKBUILD#n99

amouat commented:
ahhh, that explains a few things! Thanks @tuananh !

amouat commented:
I think we can do the setcap now as well then, I'll move this back to draft 🤦 

Apologies for the back and forth if you're following along.

amouat committed:
Set capabilities on vault binary.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

imjasonh commented on vault.yaml:
```diff
@@ -53,12 +54,31 @@ pipeline:
 
       install -m755 -D ./vault/bin/vault "${{targets.destdir}}/usr/sbin/vault"
+      setcap cap_ipc_lock=+ep  "${{targets.destdir}}/usr/sbin/vault"
+      # Binary doesn't require root, so let's move it out of sbin - for a temporary period have it
+      # in both locations. We can later delete the sbin one.
+      mkdir "${{targets.destdir}}/usr/bin"
+      chmod 755 "${{targets.destdir}}/usr/bin"
+      ln -s /usr/sbin/vault "${{targets.destdir}}/usr/bin/vault"
```
Is it in /usr/sbin/vault for compatibility with an upstream Helm chart? If so we might want to make that symlinking part of a separate compat package.

amouat replied:
The upstream Helm chart doesn't work at all with our current image. This PR adds the entrypoint package, which effectively adds Helm compatibility.

This change is separate though - I just wanted to move the binary out of sbin, as I don't think it should be there.

imjasonh replied:
I think we should just move it to `/usr/bin/vault`, if it sounds like we don't need to worry about breaking anybody since this was already not functional. WDYT?

amouat replied:
It was functional, just not compatible with the helm chart or upstream image. There was actually another breakage where the config file wasn't being picked up.

amouat replied:
I guess I'm being stupid though, that's literally a one char change to the apko that we can put through immediately.

amouat replied:
Ok, created https://github.com/chainguard-images/images/pull/731 and will resolve this conversation.

amouat commented:
I'm actually going to move this back to draft _again_.

The upstream entrypoint doesn't work very well with our package/image. I'm going to write a new entrypoint based on the upstream.

amouat committed:
Fixes for compatability with upstream.

- removed vault.hcl for compatability. It wasn't picked up in previous
  package due to typo.
- edited entrypoint to use Chainguard directory structure

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat committed:
Move binary.

 - don't worry about complexity of having 2 copies
 - also copy entrypoint to /usr/local/bin as expected by upstream.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat commented:
A few things to note about this PR:
 - the major addition is an entrypoint package to add upstream compatibility with the Helm chart
 - the vault binary was moved to /usr/bin from /usr/sbin, which requires an update to the apko
 - the config file was removed. Originally it wasn't being picked up due to a mistake in the melange config, and when it was picked up it conflicted with the entrypoint package config. The simplest solution seemed to be to remove it, as it wasn't being used.

imjasonh commented on vault.yaml:
```diff
@@ -52,13 +53,37 @@ pipeline:
-      install -m755 -D ./vault/bin/vault "${{targets.destdir}}/usr/sbin/vault"
-      install -m644 -D "./vault.hcl" "${{targets.destdir}}/etc/vault/vault.hcl"
+      install -m755 -D ./vault/bin/vault "${{targets.destdir}}/usr/bin/vault"
+      setcap cap_ipc_lock=+ep  "${{targets.destdir}}/usr/bin/vault"
```
Do we need to setcap after strip?

amouat replied:
Yes, testing that change at the moment

amouat replied:
Haven't actually got this working locally - not sure if it's because melange or apko are out of date?

amouat replied:
I've pushed the update to check if it works here...

amouat committed:
Move setcap after strip.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat committed:
Fix formatting.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat committed:
Fix permissions on directories & check for IPC_LOCK capability.

Signed-off-by: Adrian Mouat <adrian@chainguard.dev>

amouat committed:
Merge branch 'main' into vault

amouat committed:
Merge branch 'main' into vault
